modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java

```diff
@@ -0,0 +1,178 @@
+import javax.media.jai.PlanarImage;
+import javax.media.jai.ROI;
+
+import org.geotools.coverage.grid.GridCoverage2D;
+import org.geotools.coverage.grid.GridCoverageFactory;
+import org.geotools.coverage.processing.Operations;
+import org.geotools.factory.Hints;
+import org.geotools.gce.imagemosaic.GranuleDescriptor;
+import org.geotools.gce.imagemosaic.MergeBehavior;
+import org.geotools.gce.imagemosaic.MosaicElement;
+import org.geotools.gce.imagemosaic.RasterLayerRequest;
+import org.geotools.gce.imagemosaic.RasterLayerResponse;
+import org.geotools.gce.imagemosaic.RasterManager;
+import org.geotools.gce.imagemosaic.properties.CRSExtractor;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.image.ImageWorker;
+import org.geotools.referencing.CRS;
+import org.opengis.referencing.FactoryException;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+/**
+ * SubmosaicProducer that can handle reprojecting its contents into the target mosaic CRS
+ */
+class ReprojectingSubmosaicProducer extends BaseSubmosaicProducer {
+
+    private final boolean dryRun;
+
+    private final RenderingHints renderingHints;
+
+    private final Operations operations;
+
+    private CoordinateReferenceSystem targetCRS;
+
+    private RasterLayerResponse response;
+
+    private Map<String, CRSBoundMosaicProducer> perCrsMosaicProducers = new HashMap<>();
+
+    ReprojectingSubmosaicProducer(RasterLayerRequest request, RasterLayerResponse response,
+            RasterManager rasterManager, boolean dryRun) {
+        super(response, dryRun);
+        this.targetCRS = rasterManager.getConfiguration().getCrs();
+        this.response = response;
+        this.dryRun = dryRun;
+
+        Hints hints = rasterManager.getHints();
+        this.renderingHints = createRenderingHints(hints, request);
+        this.operations = new Operations(renderingHints);
+    }
+
+    private RenderingHints createRenderingHints(Hints hints, RasterLayerRequest request) {
+        RenderingHints renderHints = new RenderingHints(null);
+        if (request.getInterpolation() != null) {
+            renderingHints.put(JAI.KEY_INTERPOLATION, request.getInterpolation());
+        }
+
+        return renderHints;
+    }
+
+    @Override
+    public boolean accept(GranuleDescriptor granuleDescriptor) {
+        String granuleCRSCode = (String) granuleDescriptor.getOriginator()
+                .getAttribute(CRSExtractor.DEFAULT_ATTRIBUTE_NAME);
+        CoordinateReferenceSystem granuleCRS;
+        try {
+            granuleCRS = CRS.decode(granuleCRSCode);
+        } catch (FactoryException e) {
+            throw new IllegalStateException(e);
+        }
+
+        CRSBoundMosaicProducer submosaicProducer = this.perCrsMosaicProducers.get(granuleCRSCode);
+        if (submosaicProducer == null) {
+            submosaicProducer = new CRSBoundMosaicProducer(response, dryRun, granuleCRS);
+            perCrsMosaicProducers.put(granuleCRSCode, submosaicProducer);
```

Comment:
SortBy is respected now. I do this by grouping by every CRS + unique sort value. This could be fairly inefficient if the sort is by a large continuous range of values, but I think in many cases people will be sorting by a smaller number of discrete values (say, resolution or dates)